python/ray/ml/utils/tensorflow_utils.py

```diff
@@ -0,0 +1,56 @@
+import pandas as pd
+import tensorflow as tf
+
+
+def convert_pandas_to_tf_tensor(df: pd.DataFrame) -> tf.Tensor:
+    """Convert a pandas dataframe to a TensorFlow tensor.
+
+    This function works in two steps:
+    1. Convert each dataframe column to a tensor.
+    2. Concatenate the resulting tensors along the last axis.
+
+    Arguments:
+        df: The dataframe to convert to a TensorFlow tensor.
+
+    Returns:
+        A tensor of data type ``float32`` constructed from the dataframe.
+
+    Examples:
+        >>> import pandas as pd
+        >>> from ray.ml.utils.tensorflow_utils improt convert_pandas_to_tf_tensor
+        >>>
+        >>> df = pd.DataFrame({"X1": [1, 2, 3], "X2": [4, 5, 6]})
+        >>> convert_pandas_to_tf_tensor(df[["X1"]]).shape
+        TensorShape([3])
+        >>> convert_pandas_to_tf_tensor(df[["X1", "X2"]]).shape
+        TensorShape([3, 2])
+
+        >>> from ray.data.extensions import TensorArray
+        >>>
+        >>> df = pd.DataFrame({"image": TensorArray(np.zeros((4, 3, 32, 32)))})
+        >>> convert_pandas_to_tf_tensor(df).shape
+        TensorShape([4, 3, 32, 32])
+    """
+
+    def tensorize(series):
+        try:
+            return tf.convert_to_tensor(series)
+        except ValueError:
+            return tf.stack([tensorize(element) for element in series])
+
+    tensors = []
+    for column in df.columns:
+        series = df[column]
+        tensor = tensorize(series)
+
+        # We need to cast the tensors to a common type so that we can concatenate them.
+        # If the columns contain different types (for example, `float32`s and
+        # `int32`s), then `tf.expand_dims` raises an error.
+        tensor = tf.cast(tensor, dtype=tf.dtypes.float32)
```

Comment:
do we always want to cast to float32? it seems like the type casting should be handled by the `TypeSpec` that's passed to the TF Dataset.

Reply:
`tf.data.Dataset.from_generator` automatically casts the tensor to the appropriate type.

```
>>> def generator():
...         yield tf.range(10, dtype=tf.float32)
...
>>> for dtype in (tf.float16, tf.float64, tf.int32, tf.int64):
...         output_signature = tf.TensorSpec(shape=(10,), dtype=dtype)
...         dataset = tf.data.Dataset.from_generator(generator, output_signature=output_signature)
...         print(next(iter(dataset)))
... 
tf.Tensor([0. 1. 2. 3. 4. 5. 6. 7. 8. 9.], shape=(10,), dtype=float16)
tf.Tensor([0. 1. 2. 3. 4. 5. 6. 7. 8. 9.], shape=(10,), dtype=float64)
tf.Tensor([0 1 2 3 4 5 6 7 8 9], shape=(10,), dtype=int32)
tf.Tensor([0 1 2 3 4 5 6 7 8 9], shape=(10,), dtype=int64)
```

The choice of `float32` was somewhat arbitrary. Maybe we should use `float64` instead so users won't lose precision?

Reply:
I'm wondering if we should just let tf.cast fail in this case, or raise an error. If the user wants to convert a dataframe to a tensor, it's their responsibility to make sure the input is valid.

Would always forcing to float32 be problematic if my model expects the input to be of a different type?

Reply:
I definitely think that we should avoid this kind of silent cast, even if it will cause concatenation to fail. If the column types are incompatible, then the user should either:
1. Cast them to compatible types upstream.
2. Ensure that separate tensors are yielded for each column via the `feature_columns` arg.

One option here that could make (1) nicer is for this `convert_pandas_to_tf_tensor` function to accept an optional `tensor_spec` argument, which we can then use to perform a user-directed cast. This is pretty much what we already do for Torch tensors: https://github.com/ray-project/ray/blob/d256a31b9126a51a588ada1faa98ba9fb626d557/python/ray/ml/utils/torch_utils.py#L64

Reply:
Removed the type check for now. If we need special dtypes going forward, we can open up a PR to specify `feature_column_dtypes` similar to `Dataset.to_torch`.

Reply:
Turns out we actually do need to this, since some our tests use dataframes with multiple dtypes.

```
# From `test_to_tf`
df1 = pd.DataFrame(
        {"one": [1, 2, 3], "two": [1.0, 2.0, 3.0], "label": [1.0, 2.0, 3.0]}
    )
```

Our previous implementation used `DataFrame.values`, which implicitly casts to a common dtype.

Reply:
Oh awesome, I see that you did as such with `np.find_common_dtype()`! Nice!